pulsar/consumer_partition.go

```diff
@@ -256,6 +256,50 @@ func (pc *partitionConsumer) Close() {
 	<-req.doneCh
 }
 
+func (pc *partitionConsumer) Seek(msgID *messageID) error {
+	req := &seekRequest{
+		doneCh: make(chan struct{}),
+		msgID:  msgID,
+	}
+	pc.eventsCh <- req
+
+	// wait for the request to complete
+	<-req.doneCh
+	return req.err
+}
+
+func (pc *partitionConsumer) internalSeek(seek *seekRequest) {
+	defer close(seek.doneCh)
+
+	if pc.state == consumerClosing || pc.state == consumerClosed {
+		err := fmt.Errorf("the consumer %s was already closed when seeking the subscription %s of the topic "+
+			"%s to the message %s", pc.name, pc.options.subscription, pc.topic, seek.msgID.Serialize())
+		pc.log.WithError(err).Error("Consumer was already closed")
+		seek.err = err
+	}
+
+	id := &pb.MessageIdData{}
+	err := proto.Unmarshal(seek.msgID.Serialize(), id)
+	if err != nil {
+		pc.log.WithError(err).Errorf("deserialize message id error: %s", err.Error())
+		seek.err = err
+	}
+
+	requestID := pc.client.rpcClient.NewRequestID()
+	cmdSeek := &pb.CommandSeek{
+		ConsumerId: proto.Uint64(pc.consumerID),
+		RequestId:  proto.Uint64(requestID),
+		MessageId:  id,
+	}
+	_, err = pc.client.rpcClient.RequestOnCnx(pc.conn, requestID, pb.BaseCommand_SEEK, cmdSeek)
+	if err != nil {
+		pc.log.WithError(err).Error("Failed to reset to message id")
+		seek.err = err
+	}
+
+	pc.conn.DeleteConsumeHandler(pc.consumerID)
```

Comment:
Why are we deleting the consumer handler here? That does not seem correct.

Reply:
sure, removed.